openhcl/openhcl_boot/src/arch/x86_64/tdx.rs

```diff
@@ -94,3 +97,31 @@ impl minimal_rt::arch::IoAccess for TdxIoAccess {
         let _ = tdcall::tdcall_io_out(&mut TdcallInstruction, port, data as u32, 1);
     }
 }
+
+/// Reads MSR using TDCALL
+fn read_msr_tdcall(msr_index: u32) -> u64 {
+    let mut msr_value: u64 = 0;
+    tdcall_rdmsr(&mut TdcallInstruction, msr_index, &mut msr_value).unwrap();
+    msr_value
+}
+
+/// Global variable to store tsc frequency.
+static TSC_FREQUENCY: SingleThreaded<Cell<u64>> = SingleThreaded(Cell::new(0));
+
+/// Gets the timer ref time in 100ns, and None if it fails to get it
+pub fn get_tdx_tsc_reftime() -> Option<u64> {
+    // This is first called by the BSP from openhcl_boot and the frequency
+    // is saved in this gloabal variable. Subsequent calls use the global variable.
+    if TSC_FREQUENCY.get() == 0 {
```

Comment:
While a bit more wordy, I think `Cell<Option<u64>>` is more rusty. Although, I guess zero is an invalid value anyways? 

How can we validate this value in the future?

Reply:
I think Cell<> can be used to make the value mutable. but the caller of get_tdx_tsc_reftime() doesn't need to change the value, so I feel Option<> will suffice. On TDX-isolated VMs, if read_msr fails then TSC_FREQUENCY becomes 0, and get_tdx_tsc_reftime() returns None for future calls.

Reply:
ah i meant how we can validate the actual _value_ of TSC_FREQUENCY returned by the hypervisor. I guess our boottime values would be weird which we only use for reporting, but is there any way for the guest to validate that value?

Reply:
I guess we can validate by checking the hw-advertised frequency using CPUID which would be a TDCALL to use CPUID from openhcl_boot. I can update the TODO.TDX to use CPUID for validation.

Reply:
The hw-advertised frequency is trusted right? Should we just query that instead?

Reply:
we'll take this as an issue to followup